File: src/jray/model/ModelInstance.java
package jray.model;

import java.util.HashMap;
import java.util.Map;

import jray.common.Matrix4;
import jray.common.Object3D;
import jray.common.Ray;
import jray.common.Vect3;
import jray.math.Vect;
import jray.math.intersections.RaySphere;
import jray.struct.CollisionDetails;
import jray.struct.RayPath;

public class ModelInstance extends Object3D{

	private TriangleMeshModel model;
	private Map<Thread, CollisionData> lastCollision = new HashMap<Thread, CollisionData>();
	
	public ModelInstance(Vect3 position, TriangleMeshModel model) {
		super(position, new Vect3());
		this.model = model;
	}

	@Override
	public void rotate(Matrix4 rotationMatrix) {
		throw new RuntimeException("not implemented");
	}

	@Override
	public double getHitPointDistance(Ray r) {
		double dist;
		Ray subRay;
		Vect3 tmp = new Vect3();
		Vect.add(position, model.getBoundingSphere().getPosition(), tmp);
		
		if(RaySphere.isRayOriginatingInSphere(r.getOrigin(), r.getDirection(), tmp, model.getBoundingSphere().getRadius())){
			Vect.subtract(r.getOrigin(), position, tmp);
			subRay = new Ray(tmp, r.getDirection());
			
			dist = 0;
		}else{
			dist=RaySphere.getHitPointRaySphereDistance(r.getOrigin(), r.getDirection(), tmp, model.getBoundingSphere().getRadius());
			
			if(Double.isInfinite(dist))
				return dist;
			
			Vect.addMultiple(r.getOrigin(), r.getDirection(), dist, tmp);
			Vect.subtract(tmp, position, tmp);
			subRay = new Ray(tmp, r.getDirection());
		}
		
		CollisionData d = new CollisionData();
		lastCollision.put(Thread.currentThread(), d);
		d.d = RayPath.getFirstCollision(model.getTree(), subRay);
		if(!Double.isInfinite(d.d.d)){
			Vect3 hitPointLocal = new Vect3();
			Vect.addMultiple(subRay.getOrigin(), subRay.getDirection(), d.d.d, hitPointLocal);
			d.hitPointLocal = hitPointLocal;
			Vect3 hitPointGlobal = new Vect3();
			Vect.add(hitPointLocal, position, hitPointGlobal);
			d.hitPointGlobal = hitPointGlobal;
			return d.d.d + dist;
		}else{
			return d.d.d;
		}
	}

	@Override
	public void getNormalAt(Vect3 hitPoint, Vect3 normal) {
		CollisionData d = lastCollision.get(Thread.currentThread());
		
		if(!d.hitPointGlobal.equals(hitPoint, 1e-9))
			throw new RuntimeException("hitpoint not in cache: "+hitPoint);
		
		d.d.o.getNormalAt(d.hitPointLocal, normal);
	}

	@Override
	public boolean contains(Vect3 hitPoint) {
		throw new RuntimeException("not implemented");
	}
    
	@Override
    public Vect3 getBoundingSphereCenter(){
    	Vect3 pos = new Vect3();
    	Vect.add(position, model.getBoundingSphere().getPosition(), pos);
    	return pos;
    }
    
	@Override
    public double getBoundingSphereRadius(){
    	return model.getBoundingSphere().getRadius();
    }
}

class CollisionData{
	CollisionDetails d;
	Vect3 hitPointGlobal;
	Vect3 hitPointLocal;
}

File: JRayXLib/JRayXLib/Model/ModelInstance.cs
using System;
using System.Collections.Generic;
using System.Threading;
using JRayXLib.Math;
using JRayXLib.Math.intersections;
using JRayXLib.Shapes;
using JRayXLib.Struct;

namespace JRayXLib.Model
{
    public class ModelInstance : Basic3DObject{

        private readonly TriangleMeshModel _model;
        private readonly Dictionary<Thread, CollisionData> _lastCollision = new Dictionary<Thread, CollisionData>();

        public ModelInstance(Vect3 position, TriangleMeshModel model) : base(position, new Vect3()) {
            _model = model;
        }

        public override void Rotate(Matrix4 rotationMatrix) {
            throw new Exception("not implemented");
        }

        public override double GetHitPointDistance(Shapes.Ray r) {
            double dist;
            Shapes.Ray subRay;
            var tmp = new Vect3();
            Vect.Add(Position, _model.GetBoundingSphere().Position, tmp);
		
            if(RaySphere.IsRayOriginatingInSphere(r.GetOrigin(), r.GetDirection(), tmp, _model.GetBoundingSphere().GetRadius())){
                Vect.subtract(r.GetOrigin(), Position, tmp);
                subRay = new Shapes.Ray(tmp, r.GetDirection());
			
                dist = 0;
            }else{
                dist=RaySphere.GetHitPointRaySphereDistance(r.GetOrigin(), r.GetDirection(), tmp, _model.GetBoundingSphere().GetRadius());
			
                if(double.IsInfinity(dist))
                    return dist;
			
                Vect.AddMultiple(r.GetOrigin(), r.GetDirection(), dist, tmp);
                Vect.subtract(tmp, Position, tmp);
                subRay = new Shapes.Ray(tmp, r.GetDirection());
            }
		
            var d = new CollisionData
                {
                    Details = RayPath.GetFirstCollision(_model.GetTree(), subRay)
                };

            if(!double.IsInfinity(d.Details.Distance)){
                var hitPointLocal = new Vect3();
                Vect.AddMultiple(subRay.GetOrigin(), subRay.GetDirection(), d.Details.Distance, hitPointLocal);
                d.HitPointLocal = hitPointLocal;
                var hitPointGlobal = new Vect3();
                Vect.Add(hitPointLocal, Position, hitPointGlobal);
                d.HitPointGlobal = hitPointGlobal;
                return d.Details.Distance + dist;
            }
            _lastCollision[Thread.CurrentThread] = d;
            return d.Details.Distance;
        }

        public override void GetNormalAt(Vect3 hitPoint, Vect3 normal) {
            CollisionData d;

            if (_lastCollision.TryGetValue(Thread.CurrentThread, out d)
                && d.HitPointGlobal.Equals(hitPoint, Constants.EPS))
            {
                d.Details.Obj.GetNormalAt(d.HitPointLocal, normal);
            }

            throw new Exception("hitpoint not in cache: " + hitPoint);
        }

        public override bool Contains(Vect3 hitPoint) {
            throw new Exception("not implemented");
        }
    
        public new Vect3 GetBoundingSphereCenter(){
            var pos = new Vect3();
            Vect.Add(Position, _model.GetBoundingSphere().Position, pos);
            return pos;
        }
    
        public override double GetBoundingSphereRadius(){
            return _model.GetBoundingSphere().GetRadius();
        }
    }

    internal class CollisionData{
        internal CollisionDetails Details;
        internal Vect3 HitPointGlobal;
        internal Vect3 HitPointLocal;
    }
}